lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1349,12 +1407,48 @@ class MongoClientInterface {
      * @param {Function} cb callback
      * @return {undefined}
      */
-    internalListObject(bucketName, params, extension, log, cb) {
+    internalListObjectV1(bucketName, params, extension, log, cb) {
         const c = this.getCollection(bucketName);
+        const getLatestVersion = this.getLatestVersion;
         let stream;
         if (!params.secondaryStreamParams) {
-            stream = new MongoReadStream(c, params.mainStreamParams, params.mongifiedSearch);
+            /**
+             * When listing masters only we can't just skip PHD
+             * we have to replace them with the latest version of
+             * the object.
+             * Here we use a trasform stream that we pipe with the
+             * mongo read steam and that checks and replaces the key
+             * read if it's a PHD
+             *  */
+            const resolvePhdKey = new Transform({
+                objectMode: true,
+                transform(obj, encoding, callback) {
+                    if (Version.isPHD(obj.value)) {
+                        const key = obj.key.slice(DB_PREFIXES.Master.length);
+                        getLatestVersion(c, key, BUCKET_VERSIONS.v1, log, (err, version) => {
+                            if (err) {
+                                log.error(
+                                    'internalListObjectV1: error while getting latest version of PHD key');
+                                return callback(errors.InternalError);
+                            }
+                            MongoUtils.unserialize(version);
+                            // we keep the master key and only replace the value
+                            const latestVersion = {
+                                key: obj.key,
+                                value: JSON.stringify(version),
+                            };
+                            return callback(null, latestVersion);
+                        });
+                    } else {
+                        callback(null, obj);
+                    }
+                },
+            });
+            // listing masters only (DelimiterMaster)
+            stream = new MongoReadStream(c, params.mainStreamParams, params.mongifiedSearch)
+                .pipe(resolvePhdKey);
```

Comment:
the change is actually "just" introducing the pipe(resolvePhdKey) here for V1, right ?

so instead of duplicating, could we not just do something like:

```
            stream = new MongoReadStream(c, params.mainStreamParams, params.mongifiedSearch);
            if (use_v1) {
                const resolvePhdKey = ...
                stream = stream.pipe(resolvePhdKey);
            }
```

Reply:
IMO i think to make the code a bit easier to read it's better to separate them otherwise the function would just be too complex